### PR TITLE
Add low power support for STM32WL family

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Hi, this is for low power support on STM32WL family.

Based this on https://github.com/zephyrproject-rtos/zephyr/pull/35987, adding the extra switches for extra suspend-to-idle and soft-off and matching the WB/L4/L5 setup.

For stop modes I kept the mapping of the other STM32 devices (substate 1, 2, 3), for soft-off I've kept substate 0 a valid so `{PM_STATE_SOFT_OFF, 0, 0}` works.

Tested on a Seeed LoRa-E5 with the blinky sample, getting a nice 50uA quiescent current so I think it's working alright (there's an LDO with a ~50uA quiescent current on my board).

ppk-20210626T171915

### CODE REVIEW

File: drivers/timer/stm32_lptim_timer.c [hunk 1]
```diff
@@ -128,6 +128,10 @@ int sys_clock_driver_init(const struct device *dev)
 		    lptim_irq_handler, 0, 0);
 	irq_enable(DT_IRQN(DT_NODELABEL(lptim1)));
 
+#ifdef CONFIG_SOC_SERIES_STM32WLX
+	__HAL_LPTIM_LPTIM1_EXTI_ENABLE_IT();
```
Comment (erwango) [inline drivers/timer/stm32_lptim_timer.c:132]: I'm not clear on this addition. Why is that required only for WL ?
Also, if really needed, using LL API would be cleaner.

**Review by erwango — CHANGES_REQUESTED** [review]
One question on LPTIM adaptation, otherwise sounds good! Thanks !

File: drivers/timer/stm32_lptim_timer.c (hunk 1 above)
Comment (fabiobaltieri) [reply drivers/timer/stm32_lptim_timer.c:132]: I'm not too sure but it seems like that's how LPTIM1 wakeup works on WL series, think it has something to do with the dual core option. The EXTI channels says "LPTIMx wakeup".

x

I see that HAL function defined for other architectures though, specifically G0, L4, L5 and WB. Not sure why it's not needed for those, but the WL I have does not wake up from STOP mode without that bit enabled.

Switching to the LL API as suggested, just added a comment around it for the moment since that API has opaque defines, let me know if you think should structure it a bit more, like adding a `#define` for the line similarly to https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/counter/counter_ll_stm32_rtc.c#L52.

File: drivers/timer/stm32_lptim_timer.c [hunk 2]
```diff
@@ -129,7 +129,8 @@ int sys_clock_driver_init(const struct device *dev)
 	irq_enable(DT_IRQN(DT_NODELABEL(lptim1)));
 
 #ifdef CONFIG_SOC_SERIES_STM32WLX
-	__HAL_LPTIM_LPTIM1_EXTI_ENABLE_IT();
+	/* Enable the LPTIM1 wakeup EXTI line */
+	LL_EXTI_EnableIT_0_31(LL_EXTI_LINE_29);
```
Comment (erwango) [inline drivers/timer/stm32_lptim_timer.c]: Seems there was a fixup issue. Can you move this to first commit?

File: drivers/timer/stm32_lptim_timer.c (hunk 1 above)
Comment (erwango) [reply drivers/timer/stm32_lptim_timer.c]: ok, thanks. Let's keep it as is for now.

File: drivers/timer/stm32_lptim_timer.c (hunk 2 above)
Comment (fabiobaltieri) [reply drivers/timer/stm32_lptim_timer.c]: Oh yes, thanks for catching this!

**Review by erwango — APPROVED** [review]

**Review by ABOSTM — APPROVED** [review]
LGTM
